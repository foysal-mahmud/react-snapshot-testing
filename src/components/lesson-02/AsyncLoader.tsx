import React, { useState, useEffect } from 'react';

interface User {
  id: number;
  name: string;
  email: string;
  company: string;
}

interface AsyncLoaderProps {
  loadDelay?: number;
  shouldError?: boolean;
  initialData?: User[];
}

const AsyncLoader: React.FC<AsyncLoaderProps> = ({ 
  loadDelay = 1000, 
  shouldError = false,
  initialData 
}) => {
  const [users, setUsers] = useState<User[]>(initialData || []);
  const [loading, setLoading] = useState(!initialData);
  const [error, setError] = useState<string | null>(null);
  const [retryCount, setRetryCount] = useState(0);

  const fetchUsers = async () => {
    setLoading(true);
    setError(null);
    
    try {
      // Simulate API call delay
      await new Promise(resolve => setTimeout(resolve, loadDelay));
      
      if (shouldError && retryCount <= 1) {
        throw new Error('Network error occurred');
      }
      
      // Simulate successful API response
      const mockUsers: User[] = [
        { id: 1, name: 'Alice Johnson', email: 'alice@example.com', company: 'TechCorp' },
        { id: 2, name: 'Bob Smith', email: 'bob@example.com', company: 'DevCo' },
        { id: 3, name: 'Carol Davis', email: 'carol@example.com', company: 'CodeWorks' },
      ];
      
      setUsers(mockUsers);
      setError(null); // Clear error on success
    } catch (err) {
      setError(err instanceof Error ? err.message : 'An error occurred');
    } finally {
      setLoading(false);
    }
  };

  const handleRetry = () => {
    setRetryCount(prev => prev + 1);
    setTimeout(() => fetchUsers(), 100); // Small delay to see updated count
  };

  const handleRefresh = () => {
    setRetryCount(0);
    fetchUsers();
  };

  const handleLoadMore = async () => {
    setLoading(true);
    
    try {
      await new Promise(resolve => setTimeout(resolve, 500));
      
      const newUsers: User[] = [
        { id: users.length + 1, name: 'David Wilson', email: 'david@example.com', company: 'NewTech' },
        { id: users.length + 2, name: 'Eva Brown', email: 'eva@example.com', company: 'FutureCorp' },
      ];
      
      setUsers(prev => [...prev, ...newUsers]);
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Failed to load more users');
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    if (!initialData) {
      fetchUsers();
    }
  }, []);

  // Loading state
  if (loading && users.length === 0) {
    return (
      <div className="async-loader" data-testid="async-loader">
        <div 
          className="loading-spinner" 
          data-testid="loading-spinner"
          role="status"
          aria-label="Loading users"
        >
          <div className="spinner"></div>
          <p>Loading users...</p>
        </div>
      </div>
    );
  }

  // Error state
  if (error && users.length === 0) {
    return (
      <div className="async-loader" data-testid="async-loader">
        <div 
          className="error-state" 
          data-testid="error-state"
          role="alert"
        >
          <h3>Oops! Something went wrong</h3>
          <p>{error}</p>
          <button 
            onClick={handleRetry}
            className="btn btn-primary"
            data-testid="retry-button"
          >
            Retry ({retryCount}/3)
          </button>
        </div>
      </div>
    );
  }

  // Success state with data
  return (
    <div className="async-loader" data-testid="async-loader">
      <div className="users-header">
        <h2>Users Directory</h2>
        <button 
          onClick={handleRefresh}
          className="btn btn-secondary"
          disabled={loading}
          data-testid="refresh-button"
          aria-label="Refresh users list"
        >
          {loading ? 'Refreshing...' : 'Refresh'}
        </button>
      </div>

      {error && (
        <div 
          className="error-banner" 
          role="alert"
          data-testid="error-banner"
        >
          Error: {error}
        </div>
      )}

      <div className="users-list" role="list" aria-label="Users list">
        {users.map(user => (
          <div 
            key={user.id} 
            className="user-item"
            role="listitem"
            data-testid={`user-item-${user.id}`}
          >
            <div className="user-info">
              <h3 className="user-name">{user.name}</h3>
              <p className="user-email">{user.email}</p>
              <p className="user-company">{user.company}</p>
            </div>
            <button 
              className="btn btn-small"
              aria-label={`View details for ${user.name}`}
              data-testid={`view-details-${user.id}`}
            >
              View Details
            </button>
          </div>
        ))}
      </div>

      <div className="users-footer">
        <p>
          Showing {users.length} users
        </p>
        <button 
          onClick={handleLoadMore}
          disabled={loading}
          className="btn btn-primary"
          data-testid="load-more-button"
        >
          {loading ? 'Loading...' : 'Load More'}
        </button>
      </div>

      {loading && users.length > 0 && (
        <div 
          className="loading-overlay"
          data-testid="loading-overlay"
          role="status"
          aria-label="Loading more users"
        >
          Loading more users...
        </div>
      )}
    </div>
  );
};

export default AsyncLoader; 