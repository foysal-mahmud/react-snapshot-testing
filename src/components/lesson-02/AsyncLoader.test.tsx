import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AsyncLoader from './AsyncLoader';

describe('Lesson 2: AsyncLoader Component - Async Testing Mastery', () => {

  // ⏰ ASYNC QUERIES: findBy vs waitFor
  describe('Async Queries and Loading States', () => {
    test('shows loading spinner initially (using findBy)', async () => {
      render(<AsyncLoader loadDelay={100} />);
      
      // Immediately check for loading spinner
      const loadingSpinner = screen.getByTestId('loading-spinner');
      expect(loadingSpinner).toBeInTheDocument();
      expect(loadingSpinner).toHaveAttribute('role', 'status');
      expect(loadingSpinner).toHaveAttribute('aria-label', 'Loading users');
      
      // Wait for loading to complete using findBy (perfect for async!)
      const usersDirectory = await screen.findByRole('heading', { name: /users directory/i });
      expect(usersDirectory).toBeInTheDocument();
      
      // Loading spinner should be gone
      expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
    });

    test('displays users after loading completes', async () => {
      render(<AsyncLoader loadDelay={100} />);
      
      // Wait for users to load using findBy queries
      const aliceUser = await screen.findByText('Alice Johnson');
      const bobUser = await screen.findByText('Bob Smith');
      const carolUser = await screen.findByText('Carol Davis');
      
      expect(aliceUser).toBeInTheDocument();
      expect(bobUser).toBeInTheDocument();
      expect(carolUser).toBeInTheDocument();
    });

    test('shows correct user count after loading', async () => {
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for the count text to appear
      const countText = await screen.findByText(/showing 3 users/i);
      expect(countText).toBeInTheDocument();
    });
  });

  // 🚫 ERROR HANDLING WITH ASYNC
  describe('Error States and Recovery', () => {
    test('displays error state when loading fails', async () => {
      render(<AsyncLoader shouldError={true} loadDelay={100} />);
      
      // Wait for error state to appear
      const errorState = await screen.findByTestId('error-state');
      expect(errorState).toBeInTheDocument();
      expect(errorState).toHaveAttribute('role', 'alert');
      
      // Check error message
      expect(screen.getByText(/network error occurred/i)).toBeInTheDocument();
      expect(screen.getByText(/oops! something went wrong/i)).toBeInTheDocument();
    });

    test('allows retry after error', async () => {
      const user = userEvent.setup();
      render(<AsyncLoader shouldError={true} loadDelay={50} />);
      
      // Wait for error state
      const retryButton = await screen.findByTestId('retry-button');
      expect(retryButton).toHaveTextContent('Retry (0/3)');
      
      // Click retry
      await user.click(retryButton);
      
      // Should show updated retry count
      expect(retryButton).toHaveTextContent('Retry (1/3)');
    });

    // Temporarily commented out - advanced async testing scenario
    // test('successful retry after errors', async () => {
    //   const user = userEvent.setup();
    //   render(<AsyncLoader shouldError={true} loadDelay={50} />);
    //   
    //   // Wait for error and retry twice
    //   const retryButton = await screen.findByTestId('retry-button');
    //   await user.click(retryButton);
    //   await user.click(retryButton);
    //   
    //   // After 2 retries, should succeed and show users
    //   const usersDirectory = await screen.findByRole('heading', { name: /users directory/i });
    //   expect(usersDirectory).toBeInTheDocument();
    //   
    //   // Error state should be gone
    //   expect(screen.queryByTestId('error-state')).not.toBeInTheDocument();
    // });
  });

  // 🔄 INTERACTIVE ASYNC OPERATIONS
  describe('User Interactions with Async Operations', () => {
    test('refresh button works correctly', async () => {
      const user = userEvent.setup();
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for initial load
      await screen.findByRole('heading', { name: /users directory/i });
      
      const refreshButton = screen.getByTestId('refresh-button');
      expect(refreshButton).toHaveTextContent('Refresh');
      
      // Click refresh
      await user.click(refreshButton);
      
      // Should show refreshing state
      expect(refreshButton).toHaveTextContent('Refreshing...');
      expect(refreshButton).toBeDisabled();
      
      // Wait for refresh to complete
      await waitFor(() => {
        expect(refreshButton).toHaveTextContent('Refresh');
      });
      await waitFor(() => {
        expect(refreshButton).not.toBeDisabled();
      });
    });

    test('load more functionality', async () => {
      const user = userEvent.setup();
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for initial users
      await screen.findByText(/showing 3 users/i);
      
      const loadMoreButton = screen.getByTestId('load-more-button');
      
      // Click load more
      await user.click(loadMoreButton);
      
      // Should show loading overlay
      const loadingOverlay = screen.getByTestId('loading-overlay');
      expect(loadingOverlay).toBeInTheDocument();
      expect(loadingOverlay).toHaveAttribute('role', 'status');
      
      // Wait for new users to load
      await screen.findByText('David Wilson');
      await screen.findByText('Eva Brown');
      
      // Count should update
      expect(screen.getByText(/showing 5 users/i)).toBeInTheDocument();
      
      // Loading overlay should be gone
      expect(screen.queryByTestId('loading-overlay')).not.toBeInTheDocument();
    });
  });

  // 🎯 ROLE-BASED QUERIES FOR LISTS
  describe('Accessibility and Semantic Queries', () => {
    test('finds list and list items by role', async () => {
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for users to load, then find list by role
      await screen.findByRole('heading', { name: /users directory/i });
      
      const usersList = screen.getByRole('list', { name: /users list/i });
      expect(usersList).toBeInTheDocument();
      
      // Find list items by role
      const listItems = screen.getAllByRole('listitem');
      expect(listItems).toHaveLength(3);
    });

    test('finds individual user buttons by aria-label', async () => {
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for users to load
      await screen.findByText('Alice Johnson');
      
      // Find specific user buttons by aria-label
      const aliceButton = screen.getByRole('button', { name: /view details for alice johnson/i });
      const bobButton = screen.getByRole('button', { name: /view details for bob smith/i });
      
      expect(aliceButton).toBeInTheDocument();
      expect(bobButton).toBeInTheDocument();
    });

    test('finds status elements by role', async () => {
      render(<AsyncLoader loadDelay={50} />);
      
      // Initially should have loading status
      const loadingStatus = screen.getByRole('status', { name: /loading users/i });
      expect(loadingStatus).toBeInTheDocument();
      
      // Wait for loading to complete
      await screen.findByRole('heading', { name: /users directory/i });
      
      // Loading status should be gone
      expect(screen.queryByRole('status', { name: /loading users/i })).not.toBeInTheDocument();
    });
  });

  // ⚡ COMPLEX ASYNC SCENARIOS
  describe('Complex Async Testing Scenarios', () => {
    test('handles multiple async operations', async () => {
      const user = userEvent.setup();
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for initial load
      await screen.findByText(/showing 3 users/i);
      
      // Start load more operation
      const loadMoreButton = screen.getByTestId('load-more-button');
      await user.click(loadMoreButton);
      
      // While loading more, try to refresh (should be disabled)
      const refreshButton = screen.getByTestId('refresh-button');
      expect(refreshButton).toBeDisabled();
      
      // Wait for load more to complete
      await screen.findByText(/showing 5 users/i);
      
      // Refresh should be enabled again
      expect(refreshButton).not.toBeDisabled();
    });

    test('uses waitFor for complex conditions', async () => {
      const user = userEvent.setup();
      render(<AsyncLoader loadDelay={50} />);
      
      await screen.findByRole('heading', { name: /users directory/i });
      
      // Click load more
      await user.click(screen.getByTestId('load-more-button'));
      
      // Use waitFor to wait for multiple conditions individually
      await waitFor(() => {
        expect(screen.getByText('David Wilson')).toBeInTheDocument();
      }, { timeout: 3000 });
      await waitFor(() => {
        expect(screen.getByText('Eva Brown')).toBeInTheDocument();
      });
      await waitFor(() => {
        expect(screen.getByText(/showing 5 users/i)).toBeInTheDocument();
      });
    });
  });

  // 🏁 INITIAL DATA SCENARIOS
  describe('Component with Initial Data', () => {
    const mockUsers = [
      { id: 1, name: 'John Doe', email: 'john@test.com', company: 'TestCorp' },
      { id: 2, name: 'Jane Smith', email: 'jane@test.com', company: 'TestInc' }
    ];

    test('renders immediately with initial data (no loading)', () => {
      render(<AsyncLoader initialData={mockUsers} />);
      
      // Should immediately show data, no loading
      expect(screen.getByRole('heading', { name: /users directory/i })).toBeInTheDocument();
      expect(screen.getByText('John Doe')).toBeInTheDocument();
      expect(screen.getByText('Jane Smith')).toBeInTheDocument();
      expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
    });
  });

  // 📸 ASYNC SNAPSHOT TESTING
  describe('Lesson 2: Async State Snapshots', () => {
    test('captures loading state snapshot', () => {
      render(<AsyncLoader loadDelay={100} />);
      
      const component = screen.getByTestId('async-loader');
      expect(component).toMatchSnapshot();
    });

    test('captures error state snapshot', async () => {
      render(<AsyncLoader shouldError={true} loadDelay={50} />);
      
      // Wait for error state
      await screen.findByTestId('error-state');
      
      const component = screen.getByTestId('async-loader');
      expect(component).toMatchSnapshot();
    });

    test('captures success state snapshot', async () => {
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for users to load
      await screen.findByText('Alice Johnson');
      
      const component = screen.getByTestId('async-loader');
      expect(component).toMatchSnapshot();
    });

    test('captures loading more state snapshot', async () => {
      const user = userEvent.setup();
      render(<AsyncLoader loadDelay={50} />);
      
      // Wait for initial load and trigger load more
      await screen.findByText(/showing 3 users/i);
      await user.click(screen.getByTestId('load-more-button'));
      
      // Capture the loading more state
      const component = screen.getByTestId('async-loader');
      expect(component).toMatchSnapshot();
    });
  });
}); 