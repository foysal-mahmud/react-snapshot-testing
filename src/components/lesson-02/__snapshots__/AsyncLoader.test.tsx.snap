// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson 2: AsyncLoader Component - Async Testing Mastery Lesson 2: Async State Snapshots captures error state snapshot 1`] = `
<div
  class="async-loader"
  data-testid="async-loader"
>
  <div
    class="error-state"
    data-testid="error-state"
    role="alert"
  >
    <h3>
      Oops! Something went wrong
    </h3>
    <p>
      Network error occurred
    </p>
    <button
      class="btn btn-primary"
      data-testid="retry-button"
    >
      Retry (
      0
      /3)
    </button>
  </div>
</div>
`;

exports[`Lesson 2: AsyncLoader Component - Async Testing Mastery Lesson 2: Async State Snapshots captures loading more state snapshot 1`] = `
<div
  class="async-loader"
  data-testid="async-loader"
>
  <div
    class="users-header"
  >
    <h2>
      Users Directory
    </h2>
    <button
      aria-label="Refresh users list"
      class="btn btn-secondary"
      data-testid="refresh-button"
      disabled=""
    >
      Refreshing...
    </button>
  </div>
  <div
    aria-label="Users list"
    class="users-list"
    role="list"
  >
    <div
      class="user-item"
      data-testid="user-item-1"
      role="listitem"
    >
      <div
        class="user-info"
      >
        <h3
          class="user-name"
        >
          Alice Johnson
        </h3>
        <p
          class="user-email"
        >
          alice@example.com
        </p>
        <p
          class="user-company"
        >
          TechCorp
        </p>
      </div>
      <button
        aria-label="View details for Alice Johnson"
        class="btn btn-small"
        data-testid="view-details-1"
      >
        View Details
      </button>
    </div>
    <div
      class="user-item"
      data-testid="user-item-2"
      role="listitem"
    >
      <div
        class="user-info"
      >
        <h3
          class="user-name"
        >
          Bob Smith
        </h3>
        <p
          class="user-email"
        >
          bob@example.com
        </p>
        <p
          class="user-company"
        >
          DevCo
        </p>
      </div>
      <button
        aria-label="View details for Bob Smith"
        class="btn btn-small"
        data-testid="view-details-2"
      >
        View Details
      </button>
    </div>
    <div
      class="user-item"
      data-testid="user-item-3"
      role="listitem"
    >
      <div
        class="user-info"
      >
        <h3
          class="user-name"
        >
          Carol Davis
        </h3>
        <p
          class="user-email"
        >
          carol@example.com
        </p>
        <p
          class="user-company"
        >
          CodeWorks
        </p>
      </div>
      <button
        aria-label="View details for Carol Davis"
        class="btn btn-small"
        data-testid="view-details-3"
      >
        View Details
      </button>
    </div>
  </div>
  <div
    class="users-footer"
  >
    <p>
      Showing 
      3
       users
    </p>
    <button
      class="btn btn-primary"
      data-testid="load-more-button"
      disabled=""
    >
      Loading...
    </button>
  </div>
  <div
    aria-label="Loading more users"
    class="loading-overlay"
    data-testid="loading-overlay"
    role="status"
  >
    Loading more users...
  </div>
</div>
`;

exports[`Lesson 2: AsyncLoader Component - Async Testing Mastery Lesson 2: Async State Snapshots captures loading state snapshot 1`] = `
<div
  class="async-loader"
  data-testid="async-loader"
>
  <div
    aria-label="Loading users"
    class="loading-spinner"
    data-testid="loading-spinner"
    role="status"
  >
    <div
      class="spinner"
    />
    <p>
      Loading users...
    </p>
  </div>
</div>
`;

exports[`Lesson 2: AsyncLoader Component - Async Testing Mastery Lesson 2: Async State Snapshots captures success state snapshot 1`] = `
<div
  class="async-loader"
  data-testid="async-loader"
>
  <div
    class="users-header"
  >
    <h2>
      Users Directory
    </h2>
    <button
      aria-label="Refresh users list"
      class="btn btn-secondary"
      data-testid="refresh-button"
    >
      Refresh
    </button>
  </div>
  <div
    aria-label="Users list"
    class="users-list"
    role="list"
  >
    <div
      class="user-item"
      data-testid="user-item-1"
      role="listitem"
    >
      <div
        class="user-info"
      >
        <h3
          class="user-name"
        >
          Alice Johnson
        </h3>
        <p
          class="user-email"
        >
          alice@example.com
        </p>
        <p
          class="user-company"
        >
          TechCorp
        </p>
      </div>
      <button
        aria-label="View details for Alice Johnson"
        class="btn btn-small"
        data-testid="view-details-1"
      >
        View Details
      </button>
    </div>
    <div
      class="user-item"
      data-testid="user-item-2"
      role="listitem"
    >
      <div
        class="user-info"
      >
        <h3
          class="user-name"
        >
          Bob Smith
        </h3>
        <p
          class="user-email"
        >
          bob@example.com
        </p>
        <p
          class="user-company"
        >
          DevCo
        </p>
      </div>
      <button
        aria-label="View details for Bob Smith"
        class="btn btn-small"
        data-testid="view-details-2"
      >
        View Details
      </button>
    </div>
    <div
      class="user-item"
      data-testid="user-item-3"
      role="listitem"
    >
      <div
        class="user-info"
      >
        <h3
          class="user-name"
        >
          Carol Davis
        </h3>
        <p
          class="user-email"
        >
          carol@example.com
        </p>
        <p
          class="user-company"
        >
          CodeWorks
        </p>
      </div>
      <button
        aria-label="View details for Carol Davis"
        class="btn btn-small"
        data-testid="view-details-3"
      >
        View Details
      </button>
    </div>
  </div>
  <div
    class="users-footer"
  >
    <p>
      Showing 
      3
       users
    </p>
    <button
      class="btn btn-primary"
      data-testid="load-more-button"
    >
      Load More
    </button>
  </div>
</div>
`;
